packages/aws-cdk/lib/api/hotswap/cloudwatch-logs.ts

```diff
@@ -0,0 +1,97 @@
+import { EvaluateCloudFormationTemplate } from './evaluate-cloudformation-template';
+import { CloudWatchLogEventMonitor } from './monitor/logs-monitor';
+
+// resource types that have associated CloudWatch Log Groups that should
+// be monitored
+const includedTypes = ['AWS::StepFunctions::StateMachine', 'AWS::ECS::TaskDefinition'];
+
+// resource types that will create a cloudwatch log group
+// with a specific name if one is not provided
+const implicitTypes = ['AWS::Lambda::Function', 'AWS::CodeBuild::Project'];
+
+export async function registerCloudWatchLogGroups(
+  sdkProvider: SdkProvider,
+  stackArtifact: cxapi.CloudFormationStackArtifact,
+  assetParams: { [key: string]: string },
+  hotswapLogMonitor: CloudWatchLogEventMonitor,
+): Promise<void> {
+
+  const logGroupNames = new Array<string>();
+  const logGroupLogicalIds = new Set<string>();
+  const implicitMap = new Map<string, string>();
+  //
+  // resolve the environment, so we can substitute things like AWS::Region in CFN expressions
+  const resolvedEnv = await sdkProvider.resolveEnvironment(stackArtifact.environment);
+  // create a new SDK using the CLI credentials, because the default one will not work for new-style synthesis -
+  // it assumes the bootstrap deploy Role, which doesn't have permissions to update Lambda functions
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForWriting);
+
+  // add the sdk to the hotswapLogMonitor so that it also has the
+  // correct credentials to monitor CloudWatch log groups
+  hotswapLogMonitor.addSdk(sdk);
+  // The current resources of the Stack.
+  // We need them to figure out the physical name of a resource in case it wasn't specified by the user.
+  // We fetch it lazily, to save a service call, in case all hotswapped resources have their physical names set.
+  const listStackResources = new LazyListStackResources(sdk, stackArtifact.stackName);
+  const evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+    stackArtifact,
+    parameters: assetParams,
+    account: resolvedEnv.account,
+    region: resolvedEnv.region,
+    partition: (await sdk.currentAccount()).partition,
+    urlSuffix: sdk.getEndpointSuffix,
+    listStackResources,
+  });
+
+  const resources = await listStackResources.listStackResources();
+
+  // do a first pass at identifying all log groups
+  for (const resource of resources) {
+    if (resource.ResourceType === 'AWS::Logs::LogGroup') {
+      logGroupLogicalIds.add(resource.LogicalResourceId);
+      // track all the resources that will create a log group if one
+      // is not provided
+    } else if (implicitTypes.includes(resource.ResourceType)) {
+      const pathPart = resource.ResourceType.split('::')[1].toLowerCase();
+      implicitMap.set(resource.LogicalResourceId, pathPart);
+    }
+  }
+
+  // For each CloudWatch Log group in the template, find all resources that
+  // reference that log group and if it matches once of the included types, add it.
```

Comment:
Like I said in https://github.com/aws/aws-cdk/pull/18159#discussion_r776449279 - I think instead we should filter out logs we don't want, instead of keeping only "interesting" logs.

Reply:
I'm working on updating this.

Reply:
OK, this is updated to include all CloudWatch log groups by default and only exclude a log group if it is associated with an resource type that we exclude. Currently I am excluding `AWS::EC2::FlowLog` and `AWS::CloudTrail::Trail`.